friends shards page v2

mixmix opened the pull request:
Trello : https://trello.com/c/7g1rFpsb/46-collect-others-shards-by-peer

**Purpose** : show a higher level overview of friends shards. The main design challenge was to be able to return all of a persons shards. 

**Solution**: I started with an overview of all shards, and enabled clicking into a single shard. This was a terrible idea so I made it that when you click on any shards you click into that particular friend and can see more in depth overview of all those shards. The view I built for this is minimal as this might be where @KGibb8 does some more work.

![selection_188](https://user-images.githubusercontent.com/2665886/49269011-96c59300-f4c7-11e8-895b-bd473e402c37.jpg)
_see a high level overview of your custodianship_

![selection_187](https://user-images.githubusercontent.com/2665886/49269012-96c59300-f4c7-11e8-81dc-06835194c777.jpg)
_see details and take actions (return requested shards, and later other things)_

---

**Not done:**
- [x] testing to ensure each view live updates
  - [x] on receiving a new shard
  - [x] on change of state of a shard (request comes in, or I've just returned it)
- [ ] a message when you have no shards 
  - e.g. "You haven't been sent any shards yet"

Would like feedback about whether we should get these done before merge into development or whether we should merge so kieran can do things more easily and make a card for mix (or someone) to do these

KGibb8 commented:
First look over this @mixmix and it looks great. I don't think its misaligned with my proposal to keep forwarding and returning separate either. :+1:

KGibb8 commented on plugs/app/page/dark-crystal-index.js:
```diff
@@ -42,30 +43,27 @@ exports.create = function (api) {
     const page = h('DarkCrystal -index', { title: '/dark-crystal' }, [
-      h('h1', [ 'Dark Crystal', h('i.fa.fa-diamond') ]),
-      h('section.picker', [MINE, OTHERS].map(m => {
+      h('h1', { title: '' }, [ 'Dark Crystal', h('i.fa.fa-diamond') ]),
```
Whats the purpose of an empty title here?

ameba23 replied:
i was also wondering that

mixmix replied:
Oh this is to counter the 'title' declared on line `48`. Without this anything you hover over on the page has a `/dark-crystal` title pop up which gets really annoying.
The title on line 48 is provided to tell `hypertabs` what to display in the tabs. That thing needs a bunch of love and it's a whole can of yaks that needs shaving

KGibb8 commented on views/others-shards/index.js:
```diff
@@ -1,140 +1,162 @@
-const REPLIED = 'replied'
+const RETURNED = 'returned'
 
 function DarkCrystalFriendsIndex (opts) {
```
This is confusingly named as the file is `others-shards/index.js` and its declared [here](https://github.com/blockades/patchbay-dark-crystal/pull/18/files#diff-3c685c9bc84c3a45191cccf6e22724b5R6) as `ShardsIndex`. It has 3 different names! Which one is it?

mixmix replied:
GOOD call. Thanks for pointing out the degree of this, I hate ambiguous naming. I think the app needs to opinions.

`DarkCrystalOthersShards` it is

mixmix replied:
I've re-labelled all the things more clearly.
I've settled with the domains:
- Crystals (for your secrets area)
- Friends (for the area where you interact with friend, and shards are a subview of this)

KGibb8 commented on views/others-shards/friend/show.js:
```diff
@@ -0,0 +1,90 @@
+const REQUESTED = 'requested'
+const RETURNED = 'returned'
+
+module.exports = function DarkCrystalShowFriend (opts) {
```
Just want to make sure we don't get the names all mixed up! Here its declared as `DarkCrystalShowFriend` and its grabbed as `DarkCrystalFriendShow`. Its easier to search the code when we've got them standardised.

If I'm honest, I'd roll with keeping the file namespacing consistent, and would call this `OthersShardsFriendShow` or make a `FriendShow` component that's more 'dumb' and can be used in a similar manner in other sections of the app. For example I think some of the forwarding section is going to _look_ a bit similar while being in a different place. We could probably reuse these ideas.

KGibb8 commented on views/others-shards/friend/show.js:
```diff
@@ -0,0 +1,90 @@
+    ]
+  }))
+
+  function ShardDetailed (shard) {
```
Same again its possible this is reusable? But lets cross that bridge when we come to it.

KGibb8 commented on views/others-shards/index.js:
```diff
@@ -1,140 +1,162 @@
+    )
+  }
+
+  function Shard ({ root, receivedAt, state }) {
```
I feel like this could be passed a list of classNames here instead of state, and perhaps just a title too, that makes this component super 'dumb' making it abstractable into a component a reusable.

mixmix replied:
mmm, respectfully disagree. I like re-useable, but I think this might be a premature optimisation.
I also notice if I make that change you propose that this component does nothing really.

KGibb8 commented on views/others-shards/index.js:
```diff
@@ -1,140 +1,162 @@
+      // transform is a reduce for object. Iterator signature (acc, value, key, obj)
+      const newRecordsArray = transform(newRecords, (acc, shards, feedId) => {
+        const _shards = transform(shards, (acc, state, root) => {
+          acc.push(Object.assign({ root }, state))
```
This is how the shard is determined to be passed to the `Shard` function [above](https://github.com/blockades/patchbay-dark-crystal/pull/18/files#diff-3c5dc10568d100893cb9fae3629e6d7bR60)? A bit more pre-prep of the data means you can abstract that `Shard` function as a component.

KGibb8 commented on plugs/app/page/dark-crystal-index.js:
```diff
@@ -2,9 +2,10 @@ const nest = require('depnest')
-const DarkCrystalOthersShardsIndex = require('../../../views/others-shards/index')
+const DarkCrystalIndex = require('../../../views/my-secrets/index')
+const DarkCrystalNew = require('../../../views/my-secrets/new')
+const ShardsIndex = require('../../../views/others-shards/index')
```
See https://github.com/blockades/patchbay-dark-crystal/pull/18/files#r237831023

KGibb8 commented on plugs/app/page/dark-crystal-index.js:
```diff
@@ -74,16 +72,32 @@ exports.create = function (api) {
+
     return h('section.content', { className: computed(mode, m => m === OTHERS ? '-active' : '') }, [
-      DarkCrystalOthersShardsIndex({
+      ShardsIndex({
```
See https://github.com/blockades/patchbay-dark-crystal/pull/18/files#r237831023

KGibb8 commented on views/others-shards/friend/show.js:
```diff
@@ -0,0 +1,90 @@
+  }
+}
+
+function identity (id) { return id }
```
Is it worth setting this up as a component as its used multiple times? I know its not big but it saves on rewriting code.

KGibb8 commented on views/others-shards/index.js:
```diff
@@ -1,140 +1,162 @@
+      if (err) return console.error(err)
+
+      // transform is a reduce for object. Iterator signature (acc, value, key, obj)
+      const newRecordsArray = transform(newRecords, (acc, shards, feedId) => {
```
I really like the approach you're taking here. The more data pre-prep we do, i.e. we define a comprehensive _application state_, we can make our components and views much more granular and reusable. :+1:

KGibb8 commented on plugs/app/page/dark-crystal-show.js:
```diff
@@ -25,6 +25,12 @@ exports.create = function (api) {
+    const back = () => {
+      api.app.sync.goTo({ page: 'dark-crystal' })
+      // mix: TODO close the current tab
+      // this will be especially important if / when we make the tabs hidden in a standalone install, which I think we should
```
:+1:

KGibb8 commented on views/others-shards/friend/show.js:
```diff
@@ -0,0 +1,90 @@
+          ),
+          when(returning,
+            h('i.fa.fa-spinner.fa-pulse'),
+            h('button -primary', { 'ev-click': returnShard }, 'Return Shard')
```
I can't actually see this as I don't have any shards that have been requested, but it looks like you're rendering both of these at the same time? Meaning a button to return a shard can be pressed again even though its already been returned? Worth double checking!

ameba23 commented:
im having trouble understanding the code, so not got much to say about it. but i noticed from a comment that you're already thinking about the standalone version which is great.

its also looking great with the twinkling crystals.  only thing is that intuitively, i had expected clicking on an individual crystal icon would show me something about that shard in particular. But it shows all shards from that peer, which is actually much more useful.  Maybe we want a title on the modal with something like 'Shards from <name>' to make this clear.

mixmix commented:
thanks for the review all. I've fixed the things mentioned that I think are problems and replied to the things which I don't agree with (I think they were all minor)

I've tested this across 2 computers AND EVEN FOUND A LIVE UPDATE BUG - now fixed - so that dark-crystal will live update correctly showing requests and replies and shards and everything realtime. :boom: !

@ameba23 I initially built the thing you might expect - a _single shard view_ - but immediately hated it because it was waaaay too zoomed in. It also meant that if we wanted any bulk action there was no clear place to put it in.

mixmix commented on views/friends/index.js:
```diff
@@ -0,0 +1,179 @@
+          // mix: TODO this is really crude
+          // If we ensure invites and replies have `branch` we can sort accurately (might already have?)
+          // and make sure we're replying to the most recent request (will be more relevant with ephemeral keys)
+          // e.g. sort(thread).reverse().find(isRequest) // most recent request (but is it unanswered?)
```
NOTE TO FUTURE SELVES ( also @ameba23 @KGibb8 )

when we do ephemeral returns this is gonna need revisiting. We need to write helpers which figure out the state, like which requests (invites) were replied to, and if we've replied but there's a new requests, make sure that one is displayed and interacted with. And when we reply make sure we're replying to the right request (if there are multiple)
